File: scripts/GetFourierCoeff.py
import numpy as np
from math import pi,sqrt
			 
def GetFourierCoeff(mu,Nkx,Nky,Lx,Ly,X,Y):
	"""Function that calculates the Fourier Coefficients

	Args:
		1.mu(matrix): Information map
		2.Nkx(integer):
		3.Nky(integer):
		4.Lx(integer):
		5.Ly(integer):
		6.X(matrix):
		7.Y(matrix):

	Returns:
		1.HK(matrix):
		2.muk(matrix):

	"""	
	muk=np.matrix(np.zeros((Nkx,Nky)))

	temp=np.matrix((np.append([1],sqrt(0.5)*np.matrix(np.ones((1,Nky-1))))))
	HK=np.multiply(sqrt(Lx*Ly),temp.T*temp)

	for kx in range(0,Nkx):
		for ky in range(0,Nky):
			muk[kx,ky]=(np.matrix(np.multiply(np.multiply(mu,np.cos(kx*pi*X/Lx)),np.cos(ky*pi*Y/Ly)))).sum()/HK[ky,kx]


	return	HK,muk	

File: src/GetFourierCoeff.cpp
#include "GetFourierCoeff.h"

void GetFourierCoeff(opt1 *opt,MatrixXd X,MatrixXd Y)
{
	int a;
	float PI=3.14159265359;
	int Nkx=opt->erg.Nkx;
	int Nky=opt->erg.Nky;
	MatrixXd muk(Nkx,Nky);
	MatrixXd temp1(1 , Nkx);
	MatrixXd temp2(1 , Nky);

	temp1<<1/sqrt(0.5),MatrixXd::Ones(1,Nkx-1);
	temp2<<1/sqrt(0.5),MatrixXd::Ones(1,Nky-1);
	
	//Initializing Fourier coefficents of prior
	
	muk=MatrixXd::Zero(Nkx,Nky);

	opt->erg.HK=sqrt(opt->Lx*(opt->Ly))*((sqrt(0.5)*(temp2).transpose())*(sqrt(0.5)*(temp1) ) );
	
	cout<<(opt->erg.HK).size()<<endl;
 	
 	for(int kx=0;kx< Nkx;kx++)
 	{	
 		for(int ky=0;ky<Nky;ky++)
 		{	a=opt->erg.HK(ky,kx);
 			muk(kx,ky) = ((((opt->erg.mu).cwiseProduct((PI*(kx)*(X)/opt->Lx).array().cos().matrix()))).cwiseProduct(((PI*(ky)*(Y)/opt->Ly).array().cos().matrix())).sum())/a;
 			
 		}
 	}
 	
 	opt->erg.muk=muk;

}